File: src/cpp/multithreading/producer_consumer.cpp
/**
 * Producer-Consumer Pattern
 *
 * This example demonstrates the producer-consumer pattern using a
 * thread-safe queue. Producers add items, consumers take them.
 *
 * Key concepts:
 * - Thread-safe queue using mutex + condition_variable
 * - Producers signal consumers when items are added
 * - Consumers block waiting for items
 * - Poison pill pattern (-1) signals termination
 * - Multiple consumers can process the same queue
 */

#include <condition_variable>
#include <iostream>
#include <mutex>
#include <queue>
#include <thread>

class TaskQueue {
public:
    void add_task(int task) {
        {
            std::scoped_lock lock(mutex_);
            queue_.push(task);
        }
        cv_.notify_one();
    }

    int take_task() {
        std::unique_lock<std::mutex> lock(mutex_);
        cv_.wait(lock, [this] { return !queue_.empty(); });

        int task = queue_.front();
        queue_.pop();
        return task;
    }

private:
    std::queue<int> queue_;
    std::mutex mutex_;
    std::condition_variable cv_;
};

TaskQueue task_queue;

void producer() {
    // Add tasks
    for (int i = 0; i < 10000; ++i) {
        task_queue.add_task(1);
    }

    // Signal end with poison pill
    task_queue.add_task(-1);
    std::cout << "Producer: finished adding tasks.\n";
}

void consumer(int id) {
    int tasks_completed = 0;

    while (true) {
        int task = task_queue.take_task();

        if (task == -1) {
            // Put poison pill back for other consumers
            task_queue.add_task(-1);
            break;
        }

        tasks_completed += task;
    }

    std::cout << "Consumer " << id << ": completed " << tasks_completed
              << " tasks.\n";
}

int main() {
    std::thread prod(producer);
    std::thread cons1(consumer, 1);
    std::thread cons2(consumer, 2);

    prod.join();
    cons1.join();
    cons2.join();

    return 0;
}

File: src/python/multithreading/producer_consumer.py
"""
Producer-Consumer Pattern Example

This script demonstrates the classic producer-consumer pattern using Python threads.
Producers create items and add them to a shared buffer, while consumers take items
from the buffer and process them.

Key Concepts:
- queue.Queue provides a thread-safe FIFO queue
- Producers put() items, consumers get() items
- Queue handles all synchronization internally
- Use task_done() and join() when the main thread needs to know all queued work is done
- Use None as a sentinel value to signal consumers to stop

This pattern is fundamental for:
- Task distribution systems
- Data processing pipelines
- Event-driven architectures
- Load balancing

The Queue's maxsize parameter can be used to implement backpressure,
causing producers to block when the queue is full.
"""

import random
import time
from queue import Queue
from threading import Thread


def producer(buffer, producer_id):
    """Produce 10 random items and add them to the buffer."""
    for _ in range(10):
        item = random.randint(1, 100)
        buffer.put(item)
        print(f"Producer {producer_id} produced item {item}")
        time.sleep(random.uniform(0.1, 0.5))


def consumer(buffer, consumer_id):
    """Consume items from the buffer until receiving None sentinel."""
    while True:
        item = buffer.get()
        try:
            if item is None:  # Sentinel value signals shutdown
                break
            print(f"Consumer {consumer_id} consumed item {item}")
            time.sleep(random.uniform(0.1, 0.5))
        finally:
            # Every get() must have a matching task_done() when Queue.join() is used.
            buffer.task_done()


def main():
    buffer = Queue(maxsize=10)  # Bounded buffer with backpressure
    num_producers = 3
    num_consumers = 3

    producers = [
        Thread(target=producer, args=(buffer, i + 1)) for i in range(num_producers)
    ]
    consumers = [
        Thread(target=consumer, args=(buffer, i + 1)) for i in range(num_consumers)
    ]

    # Start all threads
    for p in producers:
        p.start()
    for c in consumers:
        c.start()

    # Wait for producers to finish
    for p in producers:
        p.join()

    # Signal consumers to stop by sending sentinel values
    for _ in range(num_consumers):
        buffer.put(None)

    # Wait until every produced item and every sentinel has been consumed
    buffer.join()

    # Wait for consumers to finish
    for c in consumers:
        c.join()


if __name__ == "__main__":
    main()
